File: apps/desktop/src/combat_meter_settings.rs
use std::{
    collections::BTreeMap,
    path::{Path, PathBuf},
};

use serde::{Deserialize, Serialize};

const SCHEMA_VERSION: u16 = 1;
const MAX_SETTINGS_BYTES: u64 = 64 * 1024;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PlayerDetailPresentation {
    InAppLayer,
    Popover,
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum HistoryPartyColorMode {
    #[default]
    PartyOrder,
    Randomized,
    Specialization,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct HistoryPartyViewSettings {
    pub id: String,
    pub label: String,
    pub columns: Vec<String>,
    #[serde(default)]
    pub widths: BTreeMap<String, u16>,
    pub sort_key: String,
    pub sort_direction: String,
    #[serde(default = "default_history_detail_mode")]
    pub detail_mode: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct CombatMeterSettings {
    pub schema_version: u16,
    pub player_detail_presentation: PlayerDetailPresentation,
    #[serde(default = "default_true")]
    pub show_class: bool,
    #[serde(default = "default_true")]
    pub show_specialization: bool,
    #[serde(default = "default_true")]
    pub show_level: bool,
    #[serde(default = "default_true")]
    pub show_ability_score: bool,
    #[serde(default = "default_true")]
    pub show_seasonal_score: bool,
    #[serde(default = "default_true")]
    pub show_character_uid: bool,
    #[serde(default = "default_true")]
    pub show_party_icons: bool,
    #[serde(default = "default_true")]
    pub show_weapon: bool,
    #[serde(default = "default_true")]
    pub show_primary_imagines: bool,
    #[serde(default = "default_true")]
    pub show_role_loadout: bool,
    #[serde(default = "default_true")]
    pub show_history_player_column: bool,
    #[serde(default = "default_true")]
    pub show_history_damage_column: bool,
    #[serde(default = "default_true")]
    pub show_history_dps_column: bool,
    #[serde(default = "default_true")]
    pub show_history_encounter_dps_column: bool,
    #[serde(default = "default_true")]
    pub show_history_hps_column: bool,
    #[serde(default = "default_true")]
    pub show_history_tps_column: bool,
    #[serde(default = "default_true")]
    pub show_history_rdps_column: bool,
    #[serde(default = "default_true")]
    pub show_history_apm_column: bool,
    #[serde(default = "default_true")]
    pub show_history_deaths_column: bool,
    #[serde(default = "default_history_party_views")]
    pub history_party_views: Vec<HistoryPartyViewSettings>,
    #[serde(default)]
    pub history_party_color_mode: HistoryPartyColorMode,
    #[serde(default)]
    pub history_specialization_colors: BTreeMap<String, String>,
    #[serde(default = "default_history_body_font_size_px")]
    pub history_body_font_size_px: u16,
    #[serde(default = "default_history_heading_font_size_px")]
    pub history_heading_font_size_px: u16,
    #[serde(default = "default_history_table_font_size_px")]
    pub history_table_font_size_px: u16,
    #[serde(default = "default_history_metadata_font_size_px")]
    pub history_metadata_font_size_px: u16,
    #[serde(default = "default_history_metric_font_size_px")]
    pub history_metric_font_size_px: u16,
    #[serde(default = "default_history_icon_size_px")]
    pub history_icon_size_px: u16,
}

const fn default_true() -> bool {
    true
}

const fn default_history_body_font_size_px() -> u16 {
    15
}

const fn default_history_heading_font_size_px() -> u16 {
    24
}

const fn default_history_table_font_size_px() -> u16 {
    13
}

const fn default_history_metadata_font_size_px() -> u16 {
    11
}

const fn default_history_metric_font_size_px() -> u16 {
    18
}

const fn default_history_icon_size_px() -> u16 {
    48
}

impl Default for CombatMeterSettings {
    fn default() -> Self {
        Self {
            schema_version: SCHEMA_VERSION,
            player_detail_presentation: PlayerDetailPresentation::InAppLayer,
            show_class: true,
            show_specialization: true,
            show_level: true,
            show_ability_score: true,
            show_seasonal_score: true,
            show_character_uid: true,
            show_party_icons: true,
            show_weapon: true,
            show_primary_imagines: true,
            show_role_loadout: true,
            show_history_player_column: true,
            show_history_damage_column: true,
            show_history_dps_column: true,
            show_history_encounter_dps_column: true,
            show_history_hps_column: true,
            show_history_tps_column: true,
            show_history_rdps_column: true,
            show_history_apm_column: true,
            show_history_deaths_column: true,
            history_party_views: default_history_party_views(),
            history_party_color_mode: HistoryPartyColorMode::PartyOrder,
            history_specialization_colors: BTreeMap::new(),
            history_body_font_size_px: default_history_body_font_size_px(),
            history_heading_font_size_px: default_history_heading_font_size_px(),
            history_table_font_size_px: default_history_table_font_size_px(),
            history_metadata_font_size_px: default_history_metadata_font_size_px(),
            history_metric_font_size_px: default_history_metric_font_size_px(),
            history_icon_size_px: default_history_icon_size_px(),
        }
    }
}

#[derive(Debug)]
pub struct CombatMeterSettingsStore {
    path: PathBuf,
    settings: CombatMeterSettings,
}

impl CombatMeterSettingsStore {
    pub fn open(path: impl Into<PathBuf>) -> Result<Self, String> {
        let path = path.into();
        let settings = load(&path)?;
        Ok(Self { path, settings })
    }

    pub fn snapshot(&self) -> CombatMeterSettings {
        self.settings.clone()
    }

    pub fn update(&mut self, settings: CombatMeterSettings) -> Result<CombatMeterSettings, String> {
        validate(&settings)?;
        write(&self.path, &settings)?;
        self.settings = settings;
        Ok(self.snapshot())
    }
}

fn load(path: &Path) -> Result<CombatMeterSettings, String> {
    let metadata = match std::fs::metadata(path) {
        Ok(metadata) => metadata,
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => {
            return Ok(CombatMeterSettings::default());
        }
        Err(error) => return Err(format!("could not inspect Combat Meter settings: {error}")),
    };
    if metadata.len() > MAX_SETTINGS_BYTES {
        return Err("Combat Meter settings exceed the 64 KiB safety limit".into());
    }
    let bytes = std::fs::read(path)
        .map_err(|error| format!("could not read Combat Meter settings: {error}"))?;
    let settings = serde_json::from_slice(&bytes)
        .map_err(|error| format!("Combat Meter settings are invalid: {error}"))?;
    validate(&settings)?;
    Ok(settings)
}

fn validate(settings: &CombatMeterSettings) -> Result<(), String> {
    if settings.schema_version != SCHEMA_VERSION {
        return Err(format!(
            "unsupported Combat Meter settings schema {}; expected {SCHEMA_VERSION}",
            settings.schema_version
        ));
    }
    if settings.history_specialization_colors.len() > 256 {
        return Err("History specialization colors exceed the 256-entry safety limit".into());
    }
    validate_history_party_views(&settings.history_party_views)?;
    for (key, color) in &settings.history_specialization_colors {
        if key.is_empty()
            || key.len() > 64
            || !key
                .bytes()
                .all(|byte| byte.is_ascii_alphanumeric() || b"._:-".contains(&byte))
        {
            return Err(format!(
                "History specialization color key {key:?} is invalid"
            ));
        }
        if color.len() != 7
            || !color.starts_with('#')
            || !color[1..].bytes().all(|byte| byte.is_ascii_hexdigit())
        {
            return Err(format!(
                "History specialization color for {key:?} is invalid"
            ));
        }
    }
    for (label, value, minimum, maximum) in [
        (
            "History body font size",
            settings.history_body_font_size_px,
            11,
            24,
        ),
        (
            "History heading font size",
            settings.history_heading_font_size_px,
            16,
            40,
        ),
        (
            "History table font size",
            settings.history_table_font_size_px,
            10,
            24,
        ),
        (
            "History metadata font size",
            settings.history_metadata_font_size_px,
            9,
            20,
        ),
        (
            "History metric font size",
            settings.history_metric_font_size_px,
            13,
            36,
        ),
        ("History icon size", settings.history_icon_size_px, 20, 64),
    ] {
        if !(minimum..=maximum).contains(&value) {
            return Err(format!(
                "{label} must be between {minimum} and {maximum} px"
            ));
        }
    }
    Ok(())
}

fn default_history_party_views() -> Vec<HistoryPartyViewSettings> {
    [
        (
            "damage",
            "Damage",
            &["player", "damage", "dps", "encounterDps", "rdps", "deaths"][..],
            "encounterDps",
        ),
        (
            "healing",
            "Healing",
            &[
                "player",
                "effectiveHealing",
                "healing",
                "shielding",
                "hps",
                "deaths",
            ][..],
            "hps",
        ),
        (
            "defense",
            "Defense",
            &["player", "damageTaken", "tps", "deaths"][..],
            "tps",
        ),
    ]
    .into_iter()
    .map(|(id, label, columns, sort_key)| HistoryPartyViewSettings {
        id: id.into(),
        label: label.into(),
        columns: columns.iter().map(|column| (*column).into()).collect(),
        widths: BTreeMap::new(),
        sort_key: sort_key.into(),
        sort_direction: "descending".into(),
        detail_mode: match id {
            "healing" => "healing",
            "defense" => "defense",
            _ => "damage",
        }
        .into(),
    })
    .collect()
}

fn validate_history_party_views(views: &[HistoryPartyViewSettings]) -> Result<(), String> {
    const COLUMNS: &[&str] = &[
        "player",
        "damage",
        "effectiveDamage",
        "damageTaken",
        "healing",
        "effectiveHealing",
        "shielding",
        "hits",
        "criticalRate",
        "dps",
        "encounterDps",
        "hps",
        "tps",
        "rdps",
        "rdpsGiven",
        "rdpsReceived",
        "apm",
        "deaths",
    ];
    if views.is_empty() || views.len() > 12 {
        return Err("History must contain between 1 and 12 party views".into());
    }
    let mut ids = std::collections::BTreeSet::new();
    for view in views {
        if view.id.is_empty()
            || view.id.len() > 40
            || !view
                .id
                .bytes()
                .all(|byte| byte.is_ascii_alphanumeric() || byte == b'_' || byte == b'-')
            || !ids.insert(view.id.as_str())
        {
            return Err(format!(
                "History party view {:?} has an invalid or duplicate ID",
                view.id
            ));
        }
        let label = view.label.trim();
        if label.is_empty() || label.len() > 32 {
            return Err(format!(
                "History party view {} has an invalid label",
                view.id
            ));
        }
        if view.columns.is_empty() || view.columns.len() > COLUMNS.len() {
            return Err(format!(
                "History party view {} has invalid columns",
                view.id
            ));
        }
        let mut columns = std::collections::BTreeSet::new();
        for column in &view.columns {
            if !COLUMNS.contains(&column.as_str()) || !columns.insert(column.as_str()) {
                return Err(format!(
                    "History party view {} contains an invalid or duplicate column",
                    view.id
                ));
            }
        }
        if !columns.contains(view.sort_key.as_str()) {
            return Err(format!(
                "History party view {} sort column must be visible",
                view.id
            ));
        }
        if view.sort_direction != "ascending" && view.sort_direction != "descending" {
            return Err(format!(
                "History party view {} sort direction is invalid",
                view.id
            ));
        }
        if !["damage", "healing", "defense"].contains(&view.detail_mode.as_str()) {
            return Err(format!(
                "History party view {} detail mode is invalid",
                view.id
            ));
        }
        for (column, width) in &view.widths {
            if !COLUMNS.contains(&column.as_str()) || !(24..=800).contains(width) {
                return Err(format!(
                    "History party view {} width for {} is invalid",
                    view.id, column
                ));
            }
        }
    }
    Ok(())
}

fn default_history_detail_mode() -> String {
    "damage".into()
}

fn write(path: &Path, settings: &CombatMeterSettings) -> Result<(), String> {
    let parent = path
        .parent()
        .ok_or_else(|| "Combat Meter settings path has no parent".to_owned())?;
    std::fs::create_dir_all(parent)
        .map_err(|error| format!("could not create Combat Meter settings folder: {error}"))?;
    let mut bytes = serde_json::to_vec_pretty(settings)
        .map_err(|error| format!("could not encode Combat Meter settings: {error}"))?;
    bytes.push(b'\n');
    std::fs::write(path, bytes)
        .map_err(|error| format!("could not write Combat Meter settings: {error}"))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn defaults_to_the_in_app_navigation_layer() {
        let settings = CombatMeterSettings::default();
        assert_eq!(
            settings.player_detail_presentation,
            PlayerDetailPresentation::InAppLayer
        );
        assert!(settings.show_class);
        assert!(settings.show_specialization);
        assert!(settings.show_level);
        assert!(settings.show_ability_score);
        assert!(settings.show_seasonal_score);
        assert!(settings.show_character_uid);
        assert!(settings.show_party_icons);
        assert!(settings.show_weapon);
        assert!(settings.show_primary_imagines);
        assert!(settings.show_role_loadout);
        assert!(settings.show_history_player_column);
        assert!(settings.show_history_damage_column);
        assert!(settings.show_history_dps_column);
        assert!(settings.show_history_encounter_dps_column);
        assert!(settings.show_history_hps_column);
        assert!(settings.show_history_tps_column);
        assert!(settings.show_history_rdps_column);
        assert!(settings.show_history_apm_column);
        assert!(settings.show_history_deaths_column);
        assert_eq!(settings.history_party_views.len(), 3);
        assert_eq!(settings.history_party_views[0].label, "Damage");
        assert_eq!(
            settings.history_party_color_mode,
            HistoryPartyColorMode::PartyOrder
        );
        assert!(settings.history_specialization_colors.is_empty());
        assert_eq!(settings.history_body_font_size_px, 15);
        assert_eq!(settings.history_heading_font_size_px, 24);
        assert_eq!(settings.history_table_font_size_px, 13);
        assert_eq!(settings.history_metadata_font_size_px, 11);
        assert_eq!(settings.history_metric_font_size_px, 18);
        assert_eq!(settings.history_icon_size_px, 48);
    }

    #[test]
    fn old_saved_settings_receive_history_sizing_defaults() {
        let settings: CombatMeterSettings = serde_json::from_str(
            r#"{
                "schemaVersion": 1,
                "playerDetailPresentation": "in_app_layer",
                "showClass": true,
                "showSpecialization": true,
                "showLevel": true,
                "showSeasonalScore": true,
                "showPartyIcons": true
            }"#,
        )
        .unwrap();
        assert_eq!(settings.history_body_font_size_px, 15);
        assert_eq!(settings.history_icon_size_px, 48);
        assert!(settings.show_ability_score);
        assert!(settings.show_character_uid);
        assert!(settings.show_weapon);
        assert!(settings.show_primary_imagines);
        assert!(settings.show_role_loadout);
        assert!(settings.show_history_player_column);
        assert!(settings.show_history_damage_column);
        assert!(settings.show_history_dps_column);
        assert!(settings.show_history_encounter_dps_column);
        assert!(settings.show_history_hps_column);
        assert!(settings.show_history_tps_column);
        assert!(settings.show_history_rdps_column);
        assert!(settings.show_history_apm_column);
        assert!(settings.show_history_deaths_column);
        assert_eq!(settings.history_party_views.len(), 3);
        assert_eq!(
            settings.history_party_color_mode,
            HistoryPartyColorMode::PartyOrder
        );
        assert!(settings.history_specialization_colors.is_empty());
        validate(&settings).unwrap();
    }

    #[test]
    fn validates_custom_specialization_colors() {
        let mut settings = CombatMeterSettings {
            history_party_color_mode: HistoryPartyColorMode::Specialization,
            ..CombatMeterSettings::default()
        };
        settings
            .history_specialization_colors
            .insert("117".into(), "#f97316".into());
        validate(&settings).unwrap();

        settings
            .history_specialization_colors
            .insert("unsafe key".into(), "orange".into());
        assert!(validate(&settings).is_err());
    }
}

File: apps/desktop/ui/src/adapters/combat-meter-settings.ts
export type PlayerDetailPresentation = "in_app_layer" | "popover";
export type HistoryPartyColorMode = "party_order" | "randomized" | "specialization";
export type HistoryPartyColumnId =
  | "player"
  | "damage"
  | "effectiveDamage"
  | "damageTaken"
  | "healing"
  | "effectiveHealing"
  | "shielding"
  | "hits"
  | "criticalRate"
  | "dps"
  | "encounterDps"
  | "hps"
  | "tps"
  | "rdps"
  | "rdpsGiven"
  | "rdpsReceived"
  | "apm"
  | "deaths";
export type HistoryPartySortDirection = "ascending" | "descending";
export type HistoryPlayerDetailMode = "damage" | "healing" | "defense";

export interface HistoryPartyViewSettings {
  id: string;
  label: string;
  columns: HistoryPartyColumnId[];
  widths: Partial<Record<HistoryPartyColumnId, number>>;
  sortKey: HistoryPartyColumnId;
  sortDirection: HistoryPartySortDirection;
  detailMode: HistoryPlayerDetailMode;
}

export const HISTORY_PARTY_COLUMN_IDS: readonly HistoryPartyColumnId[] = [
  "player", "damage", "effectiveDamage", "damageTaken", "healing",
  "effectiveHealing", "shielding", "hits", "criticalRate", "dps",
  "encounterDps", "hps", "tps", "rdps", "rdpsGiven", "rdpsReceived",
  "apm", "deaths",
] as const;

export const DEFAULT_HISTORY_PARTY_VIEWS: readonly HistoryPartyViewSettings[] = [
  {
    id: "damage",
    label: "Damage",
    columns: ["player", "damage", "dps", "encounterDps", "rdps", "deaths"],
    widths: { player: 360, damage: 120, dps: 105, encounterDps: 105, rdps: 105, deaths: 82 },
    sortKey: "encounterDps",
    sortDirection: "descending",
    detailMode: "damage",
  },
  {
    id: "rdps",
    label: "rDPS",
    columns: ["player", "damage", "rdps", "rdpsGiven", "rdpsReceived"],
    widths: {
      player: 360,
      damage: 120,
      rdps: 110,
      rdpsGiven: 130,
      rdpsReceived: 135,
    },
    sortKey: "rdps",
    sortDirection: "descending",
    detailMode: "damage",
  },
  {
    id: "healing",
    label: "Healing",
    columns: ["player", "effectiveHealing", "healing", "shielding", "hps", "deaths"],
    widths: { player: 360, effectiveHealing: 135, healing: 120, shielding: 110, hps: 105, deaths: 82 },
    sortKey: "hps",
    sortDirection: "descending",
    detailMode: "healing",
  },
  {
    id: "defense",
    label: "Defense",
    columns: ["player", "damageTaken", "tps", "deaths"],
    widths: { player: 360, damageTaken: 135, tps: 105, deaths: 82 },
    sortKey: "tps",
    sortDirection: "descending",
    detailMode: "defense",
  },
] as const;

export const HISTORY_PARTY_PALETTE = [
  "#5eead4", "#60a5fa", "#f472b6", "#fbbf24", "#a78bfa",
  "#4ade80", "#fb7185", "#22d3ee", "#f97316", "#c084fc",
  "#2dd4bf", "#e879f9", "#84cc16", "#38bdf8", "#facc15",
  "#34d399", "#818cf8", "#f43f5e", "#14b8a6", "#d946ef",
] as const;

export function historySeededPaletteColor(seed: string, index: number): string {
  const steps = [1, 3, 7, 9, 11, 13, 17, 19] as const;
  const hash = historyColorHash(seed);
  const offset = hash % HISTORY_PARTY_PALETTE.length;
  const step = steps[(hash >>> 8) % steps.length]!;
  return HISTORY_PARTY_PALETTE[(offset + Math.max(0, index) * step) % HISTORY_PARTY_PALETTE.length]!;
}

export function historySpecializationFallbackColor(specializationId: string | number): string {
  return historySeededPaletteColor(`specialization:${specializationId}`, 0);
}

function historyColorHash(value: string): number {
  let hash = 2_166_136_261;
  for (let index = 0; index < value.length; index += 1) {
    hash ^= value.charCodeAt(index);
    hash = Math.imul(hash, 16_777_619);
  }
  return hash >>> 0;
}

export interface CombatMeterSettings {
  schemaVersion: 1;
  playerDetailPresentation: PlayerDetailPresentation;
  showClass: boolean;
  showSpecialization: boolean;
  showLevel: boolean;
  showAbilityScore: boolean;
  showSeasonalScore: boolean;
  showCharacterUid: boolean;
  showPartyIcons: boolean;
  showWeapon: boolean;
  showPrimaryImagines: boolean;
  showRoleLoadout: boolean;
  showHistoryPlayerColumn: boolean;
  showHistoryDamageColumn: boolean;
  showHistoryDpsColumn: boolean;
  showHistoryEncounterDpsColumn: boolean;
  showHistoryHpsColumn: boolean;
  showHistoryTpsColumn: boolean;
  showHistoryRdpsColumn: boolean;
  showHistoryApmColumn: boolean;
  showHistoryDeathsColumn: boolean;
  historyPartyViews: HistoryPartyViewSettings[];
  historyPartyColorMode: HistoryPartyColorMode;
  historySpecializationColors: Record<string, string>;
  historyBodyFontSizePx: number;
  historyHeadingFontSizePx: number;
  historyTableFontSizePx: number;
  historyMetadataFontSizePx: number;
  historyMetricFontSizePx: number;
  historyIconSizePx: number;
}

export const DEFAULT_COMBAT_METER_SETTINGS: CombatMeterSettings = {
  schemaVersion: 1,
  playerDetailPresentation: "in_app_layer",
  showClass: true,
  showSpecialization: true,
  showLevel: true,
  showAbilityScore: true,
  showSeasonalScore: true,
  showCharacterUid: true,
  showPartyIcons: true,
  showWeapon: true,
  showPrimaryImagines: true,
  showRoleLoadout: true,
  showHistoryPlayerColumn: true,
  showHistoryDamageColumn: true,
  showHistoryDpsColumn: true,
  showHistoryEncounterDpsColumn: true,
  showHistoryHpsColumn: true,
  showHistoryTpsColumn: true,
  showHistoryRdpsColumn: true,
  showHistoryApmColumn: true,
  showHistoryDeathsColumn: true,
  historyPartyViews: DEFAULT_HISTORY_PARTY_VIEWS.map(cloneHistoryPartyView),
  historyPartyColorMode: "party_order",
  historySpecializationColors: {},
  historyBodyFontSizePx: 15,
  historyHeadingFontSizePx: 24,
  historyTableFontSizePx: 13,
  historyMetadataFontSizePx: 11,
  historyMetricFontSizePx: 18,
  historyIconSizePx: 48,
};

export function parseCombatMeterSettings(value: unknown): CombatMeterSettings {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new Error("Combat Meter settings must be an object.");
  }
  const settings = value as Record<string, unknown>;
  if (settings.schemaVersion !== 1) {
    throw new Error("Combat Meter settings use an unsupported schema.");
  }
  if (
    settings.playerDetailPresentation !== "in_app_layer" &&
    settings.playerDetailPresentation !== "popover"
  ) {
    throw new Error("Combat Meter player detail presentation is invalid.");
  }
  if (settings.showAbilityScore === undefined) {
    settings.showAbilityScore = DEFAULT_COMBAT_METER_SETTINGS.showAbilityScore;
  }
  if (settings.showCharacterUid === undefined) {
    settings.showCharacterUid = DEFAULT_COMBAT_METER_SETTINGS.showCharacterUid;
  }
  if (settings.showWeapon === undefined) {
    settings.showWeapon = DEFAULT_COMBAT_METER_SETTINGS.showWeapon;
  }
  if (settings.showPrimaryImagines === undefined) {
    settings.showPrimaryImagines = DEFAULT_COMBAT_METER_SETTINGS.showPrimaryImagines;
  }
  if (settings.showRoleLoadout === undefined) {
    settings.showRoleLoadout = DEFAULT_COMBAT_METER_SETTINGS.showRoleLoadout;
  }
  for (const key of [
    "showHistoryPlayerColumn",
    "showHistoryDamageColumn",
    "showHistoryDpsColumn",
    "showHistoryEncounterDpsColumn",
    "showHistoryHpsColumn",
    "showHistoryTpsColumn",
    "showHistoryRdpsColumn",
    "showHistoryApmColumn",
    "showHistoryDeathsColumn",
  ] as const) {
    if (settings[key] === undefined) settings[key] = DEFAULT_COMBAT_METER_SETTINGS[key];
  }
  if (settings.historyPartyViews === undefined) {
    settings.historyPartyViews = legacyHistoryPartyViews(settings);
  }
  settings.historyPartyViews = parseHistoryPartyViews(settings.historyPartyViews);
  for (const [key, label] of [
    ["showClass", "class visibility"],
    ["showSpecialization", "specialization visibility"],
    ["showLevel", "level visibility"],
    ["showAbilityScore", "ability score visibility"],
    ["showSeasonalScore", "seasonal score visibility"],
    ["showCharacterUid", "character UID visibility"],
    ["showPartyIcons", "party icon visibility"],
    ["showWeapon", "weapon visibility"],
    ["showPrimaryImagines", "primary Imagine visibility"],
    ["showRoleLoadout", "role loadout visibility"],
    ["showHistoryPlayerColumn", "History Player column visibility"],
    ["showHistoryDamageColumn", "History Damage column visibility"],
    ["showHistoryDpsColumn", "History DPS column visibility"],
    ["showHistoryEncounterDpsColumn", "History eDPS column visibility"],
    ["showHistoryHpsColumn", "History HPS column visibility"],
    ["showHistoryTpsColumn", "History TPS column visibility"],
    ["showHistoryRdpsColumn", "History rDPS column visibility"],
    ["showHistoryApmColumn", "History APM column visibility"],
    ["showHistoryDeathsColumn", "History Deaths column visibility"],
  ] as const) {
    if (typeof settings[key] !== "boolean") {
      throw new Error(`Combat Meter ${label} is invalid.`);
    }
  }
  if (settings.historyPartyColorMode === undefined) {
    settings.historyPartyColorMode = DEFAULT_COMBAT_METER_SETTINGS.historyPartyColorMode;
  }
  if (
    settings.historyPartyColorMode !== "party_order" &&
    settings.historyPartyColorMode !== "randomized" &&
    settings.historyPartyColorMode !== "specialization"
  ) {
    throw new Error("Combat Meter History party color mode is invalid.");
  }
  if (settings.historySpecializationColors === undefined) {
    settings.historySpecializationColors = {};
  }
  if (
    typeof settings.historySpecializationColors !== "object" ||
    settings.historySpecializationColors === null ||
    Array.isArray(settings.historySpecializationColors)
  ) {
    throw new Error("Combat Meter History specialization colors are invalid.");
  }
  const specializationColors = Object.entries(
    settings.historySpecializationColors as Record<string, unknown>,
  );
  if (specializationColors.length > 256) {
    throw new Error("Combat Meter History specialization colors exceed 256 entries.");
  }
  for (const [key, color] of specializationColors) {
    if (!/^[A-Za-z0-9._:-]{1,64}$/.test(key) || typeof color !== "string" || !/^#[0-9A-Fa-f]{6}$/.test(color)) {
      throw new Error(`Combat Meter History specialization color ${key || "key"} is invalid.`);
    }
  }
  settings.historySpecializationColors = Object.fromEntries(
    specializationColors.map(([key, color]) => [key, (color as string).toLowerCase()]),
  );
  for (const [key, label, minimum, maximum] of [
    ["historyBodyFontSizePx", "History body font size", 11, 24],
    ["historyHeadingFontSizePx", "History heading font size", 16, 40],
    ["historyTableFontSizePx", "History table font size", 10, 24],
    ["historyMetadataFontSizePx", "History metadata font size", 9, 20],
    ["historyMetricFontSizePx", "History metric font size", 13, 36],
    ["historyIconSizePx", "History icon size", 20, 64],
  ] as const) {
    if (settings[key] === undefined) settings[key] = DEFAULT_COMBAT_METER_SETTINGS[key];
    const value = settings[key];
    if (typeof value !== "number" || !Number.isInteger(value) || value < minimum || value > maximum) {
      throw new Error(`${label} must be between ${minimum} and ${maximum} px.`);
    }
  }
  return settings as unknown as CombatMeterSettings;
}

export function cloneHistoryPartyView(view: HistoryPartyViewSettings): HistoryPartyViewSettings {
  return {
    ...view,
    columns: [...view.columns],
    widths: { ...view.widths },
  };
}

function legacyHistoryPartyViews(settings: Record<string, unknown>): HistoryPartyViewSettings[] {
  const legacyKeys: ReadonlyArray<readonly [HistoryPartyColumnId, string]> = [
    ["player", "showHistoryPlayerColumn"],
    ["damage", "showHistoryDamageColumn"],
    ["dps", "showHistoryDpsColumn"],
    ["encounterDps", "showHistoryEncounterDpsColumn"],
    ["hps", "showHistoryHpsColumn"],
    ["tps", "showHistoryTpsColumn"],
    ["rdps", "showHistoryRdpsColumn"],
    ["apm", "showHistoryApmColumn"],
    ["deaths", "showHistoryDeathsColumn"],
  ];
  const columns = legacyKeys
    .filter(([, key]) => settings[key] !== false)
    .map(([column]) => column);
  const view = cloneHistoryPartyView(DEFAULT_HISTORY_PARTY_VIEWS[0]!);
  view.columns = columns.length > 0 ? columns : ["player"];
  if (!view.columns.includes(view.sortKey)) {
    view.sortKey = view.columns[0]!;
    view.sortDirection = view.sortKey === "player" ? "ascending" : "descending";
  }
  return [view, ...DEFAULT_HISTORY_PARTY_VIEWS.slice(1).map(cloneHistoryPartyView)];
}

function parseHistoryPartyViews(value: unknown): HistoryPartyViewSettings[] {
  if (!Array.isArray(value) || value.length < 1 || value.length > 12) {
    throw new Error("Combat Meter History must contain between 1 and 12 party views.");
  }
  const ids = new Set<string>();
  return value.map((candidate, viewIndex) => {
    if (typeof candidate !== "object" || candidate === null || Array.isArray(candidate)) {
      throw new Error(`Combat Meter History party view ${viewIndex + 1} is invalid.`);
    }
    const view = candidate as Record<string, unknown>;
    if (typeof view.id !== "string" || !/^[A-Za-z0-9_-]{1,40}$/.test(view.id) || ids.has(view.id)) {
      throw new Error(`Combat Meter History party view ${viewIndex + 1} has an invalid or duplicate ID.`);
    }
    ids.add(view.id);
    if (typeof view.label !== "string" || view.label.trim().length < 1 || view.label.trim().length > 32) {
      throw new Error(`Combat Meter History party view ${view.id} has an invalid label.`);
    }
    if (!Array.isArray(view.columns) || view.columns.length < 1 || view.columns.length > HISTORY_PARTY_COLUMN_IDS.length) {
      throw new Error(`Combat Meter History party view ${view.id} has invalid columns.`);
    }
    const columns = view.columns as unknown[];
    const columnSet = new Set<HistoryPartyColumnId>();
    for (const column of columns) {
      if (typeof column !== "string" || !HISTORY_PARTY_COLUMN_IDS.includes(column as HistoryPartyColumnId) || columnSet.has(column as HistoryPartyColumnId)) {
        throw new Error(`Combat Meter History party view ${view.id} contains an invalid or duplicate column.`);
      }
      columnSet.add(column as HistoryPartyColumnId);
    }
    if (typeof view.sortKey !== "string" || !columnSet.has(view.sortKey as HistoryPartyColumnId)) {
      throw new Error(`Combat Meter History party view ${view.id} sort column must be visible.`);
    }
    if (view.sortDirection !== "ascending" && view.sortDirection !== "descending") {
      throw new Error(`Combat Meter History party view ${view.id} sort direction is invalid.`);
    }
    const detailMode = view.detailMode ?? (
      view.id === "healing" ? "healing" : view.id === "defense" ? "defense" : "damage"
    );
    if (detailMode !== "damage" && detailMode !== "healing" && detailMode !== "defense") {
      throw new Error(`Combat Meter History party view ${view.id} detail mode is invalid.`);
    }
    const widthsValue = view.widths ?? {};
    if (typeof widthsValue !== "object" || widthsValue === null || Array.isArray(widthsValue)) {
      throw new Error(`Combat Meter History party view ${view.id} widths are invalid.`);
    }
    const widths: Partial<Record<HistoryPartyColumnId, number>> = {};
    for (const [column, width] of Object.entries(widthsValue as Record<string, unknown>)) {
      if (!HISTORY_PARTY_COLUMN_IDS.includes(column as HistoryPartyColumnId) || typeof width !== "number" || !Number.isInteger(width) || width < 24 || width > 800) {
        throw new Error(`Combat Meter History party view ${view.id} width for ${column} is invalid.`);
      }
      widths[column as HistoryPartyColumnId] = width;
    }
    return {
      id: view.id,
      label: view.label.trim(),
      columns: [...columnSet],
      widths,
      sortKey: view.sortKey as HistoryPartyColumnId,
      sortDirection: view.sortDirection,
      detailMode,
    };
  });
}
